src/mlpack/methods/ann/rbm_impl.hpp

```diff
@@ -0,0 +1,241 @@
+    const arma::Mat<ElemType>& /* parameters*/,
+    const size_t i,
+    const size_t batchSize)
+{
+  if (!useMonitoringCost)
+  {
+    Gibbs(std::move(predictors.cols(i, i + batchSize - 1)),
+          std::move(negativeSamples));
+    return std::fabs(FreeEnergy(std::move(predictors.cols(i,
+        i + batchSize - 1))) - FreeEnergy(std::move(negativeSamples)));
+  }
+  else
+  {
+    if (persistence)
+    {
+      return rbmPolicy.Evaluate(predictors, i, batchSize);
+    }
+    else
+    {
+      // Mean Squared Error
+      rbmPolicy.SampleHidden(std::move(predictors.cols(i, i + batchSize - 1)),
+          std::move(hiddenReconstruction));
+      rbmPolicy.SampleVisible(std::move(hiddenReconstruction),
+          std::move(visibleReconstruction));
+      return arma::accu(arma::pow(visibleReconstruction -
+                                  predictors.cols(i, i + batchSize - 1),
+          2)) / hiddenReconstruction.n_rows;
+    }
+  }
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::SampleHidden(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output)
+{
+  rbmPolicy.SampleHidden(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::SampleVisible(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output)
+{
+  rbmPolicy.SampleVisible(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Gibbs(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output,
+    size_t steps)
+{
+  steps = (steps == SIZE_MAX) ? this-> numSteps: steps;
+
+  if (persistence && !state.is_empty())
+  {
+    SampleHidden(std::move(state), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+  else
+  {
+    SampleHidden(std::move(input), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+
+  for (size_t j = 1; j < steps; j++)
+  {
+    SampleHidden(std::move(output), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+  if (persistence)
+    state = output;
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Gradient(
+    arma::Mat<ElemType>& /*parameters*/,
+    const size_t input,
+    arma::Mat<ElemType>& output,
+    const size_t batchSize)
+{
+  positiveGradient.zeros();
+  // Collect the negative samples
+  rbmPolicy.PositivePhase(std::move(predictors.cols(input,
+                                                    input + batchSize - 1)),
+                          std::move(positiveGradient));
+
+  negativeGradient.zeros();
+
+  for (size_t i = 0; i < negSteps; i++)
+  {
+    Gibbs(std::move(predictors.cols(input, input + batchSize - 1)),
```

Comment:
`Looks like, some matrix dimensions are incorrect`.
Actually, this was because here, I tried to directly pass a batch of columns, which is wrong because internally, the present code expects input as a single column in one pass. For, now I can just introduce a `for` loop over the `batchSize`, and that should work. Do we want to leave that as it is, or should the implementation incorporate the use case of input being a bunch of columns rather than a single column? cc @zoq

Reply:
hm, you should be able to pass a matrix where each column is another input, at least this is what we do for the FFN class and we do something similar for the RNN class, which uses arma::cube but just to handle data sequences.The format of the data should be as follows:

- each slice should correspond to a time step
- each column should correspond to a data point
- each row should correspond to a dimension

anyway, you are right that in case of a single input which is multidimensional it's represented as a single column, maybe it's worth to use ``arma::cube`` here as well. Let me know what you think.

Reply:
It sounds logical to me, I'll have to change the implementation a bit. @lozhnikov What do you prefer?

Reply:
Ping @lozhnikov.

Reply:
@ShikharJ Sorry for the slow response, I was preparing a report on gas dynamics. I agree with Marcus. The idea looks good to me.